File: app/src/main/java/com/corona/covid_19tracker/Activitys/Fragment/BDdataActivity.java
package com.corona.covid_19tracker.Activitys.Fragment;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.corona.covid_19tracker.Adapter.BD_DisRecyclerAdapter;
import com.corona.covid_19tracker.Model.Bd_dis_item;
import com.corona.covid_19tracker.DoInBackground.doInBackground;
import com.corona.covid_19tracker.LoadingDialog;
import com.corona.covid_19tracker.R;
import java.util.ArrayList;


public class BDdataActivity extends Fragment {
    private ArrayList<Bd_dis_item> disData;
    private RecyclerView recyclerView;
    private Dialog loadingDialog;
    private BD_DisRecyclerAdapter mAdapter;
    private boolean flag = true;
    private  View view = null;
    private SwipeRefreshLayout swipeRefreshLayout;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

         view = inflater.inflate(R.layout.activity_bddata, container, false);
         swipeRefreshLayout = view.findViewById(R.id.swapLayout);
      //  loading dialog
        loadingDialog = LoadingDialog.getCustomLoadingDialog(getContext());
        loadingDialog.show();

        //recycler
        recyclerView = view.findViewById(R.id.bdRecycler);

        LinearLayoutManager llm = new LinearLayoutManager(getContext());
        llm.setOrientation(LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(llm);
        recyclerView.setHasFixedSize(true);


        getData();


        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getData();
            }
        });


        return view;
    }

    private void getData() {
        disData = new ArrayList<>();
        doInBackground bdDisData = new doInBackground(getContext());
        bdDisData.districtDataService(new doInBackground.VolleyDisResponseListener() {
            @Override
            public void onResponse(ArrayList<Bd_dis_item> dataBd) {
                disData = dataBd;
                flag = true;
                loadingDialog.dismiss();
                flag = true;
                setRecyclerAdapter();
            }

            @Override
            public void onErrorResponse() {
                if(flag) {
                    Toast.makeText(getContext(), "Check Your Internet Connection", Toast.LENGTH_SHORT).show();
                    flag = false;
                }
                getData();
                swipeRefreshLayout.setRefreshing(false);
            }
        });
    }

    //set adapter on recyclerView
    private void setRecyclerAdapter() {
        mAdapter = new BD_DisRecyclerAdapter(disData, getContext());
        recyclerView.setAdapter(mAdapter);
        loadingDialog.dismiss();
        swipeRefreshLayout.setRefreshing(false);
        mAdapter.notifyDataSetChanged();
    }


}

File: app/src/main/java/com/corona/covid_19tracker/Activity/Fragment/BDdataActivity.kt
package com.corona.covid_19tracker.Activity.Fragment

import android.app.Dialog
import androidx.recyclerview.widget.RecyclerView
import com.corona.covid_19tracker.Adapter.BD_DisRecyclerAdapter
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.corona.covid_19tracker.Api.BdDataRetrofitHelper
import com.corona.covid_19tracker.Api.DistrictDataService
import com.corona.covid_19tracker.LoadingDialog
import com.corona.covid_19tracker.Model.DistrictDataModel
import com.corona.covid_19tracker.Repository.BDdataRepository
import com.corona.covid_19tracker.Repository.Response
import com.corona.covid_19tracker.ViewModel.BDdataActivityViewModel
import com.corona.covid_19tracker.ViewModel.BDdataActivityViewModelFactory
import com.corona.covid_19tracker.databinding.ActivityBddataBinding

class BDdataActivity : Fragment() {
    private var recyclerView: RecyclerView? = null
    private var loadingDialog: Dialog? = null
    private var mAdapter: BD_DisRecyclerAdapter? = null
    private val flag = true
    private lateinit var binding: ActivityBddataBinding
    private var swipeRefreshLayout: SwipeRefreshLayout? = null
    private var counter = 1
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = ActivityBddataBinding.inflate(layoutInflater)
        binding.shimmerViewContainer.stopShimmer()

        val BdStateDataService =
            BdDataRetrofitHelper.getInstance().create(DistrictDataService::class.java)
        val repository = BDdataRepository(BdStateDataService)
        val viewModel =
            ViewModelProvider(requireActivity(), BDdataActivityViewModelFactory(repository)).get(
                BDdataActivityViewModel::class.java
            )
        viewModel.result.observe(requireActivity(), Observer {
            it.let {
                when (it) {
                    is Response.Loading -> {

                    }
                    is Response.Success -> {
                        setData(it.data!!)
                    }
                    is Response.Error -> {
                        viewModel.getResult()
                        if (counter == 1)
                            Toast.makeText(requireContext(), it.errorMessage, Toast.LENGTH_SHORT)
                                .show()
                        viewModel.getResult()
                        ++counter
                    }
                }
            }
        })

        return binding.root
    }

    private fun setData(districtDataModel: DistrictDataModel) {
        val llm = LinearLayoutManager(context)
        llm.orientation = LinearLayoutManager.VERTICAL
        binding.bdRecycler.layoutManager = llm
        binding.bdRecycler.setHasFixedSize(true)
        binding.bdRecycler.adapter = BD_DisRecyclerAdapter(districtDataModel)
        binding.bdRecycler.scheduleLayoutAnimation()
        binding.shimmerViewContainer.stopShimmer()
        binding.shimmerViewContainer.visibility = View.INVISIBLE
    }


}